doc/source/development/contributing.rst

```diff
@@ -696,6 +696,107 @@ You'll also need to
 
 See :ref:`contributing.warnings` for more.
 
+Type Hints
+----------
+
+*pandas* strongly encourages the use of :pep:`484` style type hints. New development should contain type hints and pull requests to annotate existing code are accepted as well!
+
+Syntax Requirements
+~~~~~~~~~~~~~~~~~~~
+
+Because *pandas* still supports Python 3.5, :pep:`526` does not apply and variables **must** be annotated with type comments. Specifically, this is a valid annotation within pandas:
+
+.. code-block:: python
+
+   primes = []  # type: List[int]
+
+Whereas this is **NOT** allowed:
+
+.. code-block:: python
+
+   primes: List[int] = []  # not supported in Python 3.5!
+
+Note that function signatures can always be annotated per :pep:`3107`:
+
+.. code-block:: python
+
+   def sum_of_primes(primes: List[int] = [])  -> int:
+
+Style Guidelines
+~~~~~~~~~~~~~~~~
+
+Types imports should follow the ``from typing import ...`` convention. So rather than
+
+.. code-block:: python
+
+   import typing
+
+   primes = []  # type: typing.List[int]
+
+You should write
+
+.. code-block:: python
+
+   from typing import List
+
+   primes = []  # type: List[int]
+
+``Optional`` should be used where applicable, so instead of
+
+.. code-block:: python
+
+   from typing import Union
+
+   maybe_primes = []  # type: Union[int, None]
+
+You should write
+
+.. code-block:: python
+
+   from typing import Optional
+
+   maybe_primes = []  # type: Optional[int]
+
+If a function accepts multiple arguments, every parameter should appear on a separate line. So rather than
+
+.. code-block:: python
+
+   def some_func(a: str, b: float, c: Union[int, float]) -> float:
+
+The preferred style would be
+
+.. code-block:: python
+
+   def some_func(a: str,
```

Comment:
I don't think we should make style recommendations that aren't enforced by a linter,

Black formats this as


```python
def some_func(a: str, b: float, c: Union[int, float]) -> float:
    pass
```

And when a wrap is required

```
def some_func(
    a_long_name_really_long_name: str,
    b_long_name: float,
    c_long_name: Union[int, float],
) -> float:
    pass
```

Reply:
Good to know. Haven't used that yet (have seen other issue, will look into it) but what you are saying makes sense.

@topper-123 I think you suggested something like this as well so FYI

Reply:
If it fits within 79 chars, I don't the point to wrap.

Reply:
I'm +1 on requiring multiline defs to be wrapped in the style guide. It makes for much clearer code.

OTOH, if we go with Black, such choices will be made for us?